File: jtms/go/jrules.go
// JTRE: A JTMS-based Tiny Rule Engine.
// Rule system module.
// Translated from jrules.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.

package jtms

import (
	"fmt"
	"io"
	"os"
)

// JRule represents a rule in the JTRE system.
// Corresponds to the Lisp defstruct rule in jrules.lisp.
type JRule struct {
	ID      int
	Jtre    *Jtre
	Dbclass *JDbclass
	Matcher func(p interface{}) (bool, []interface{}, bool)
	Body    func(args ...interface{})
}

// String implements fmt.Stringer for JRule.
func (r *JRule) String() string {
	return fmt.Sprintf("<Rule %d>", r.ID)
}

// InsertJRule creates and registers a new rule in the given dbclass.
// The rule's matcher is tried against all existing facts in the dbclass.
// Corresponds to the Lisp function insert-rule.
func InsertJRule(dbclass *JDbclass, matcher func(p interface{}) (bool, []interface{}, bool), body func(args ...interface{})) *JRule {
	jtre := dbclass.Jtre
	jtre.RuleCounter++
	rule := &JRule{
		ID:      jtre.RuleCounter,
		Jtre:    jtre,
		Dbclass: dbclass,
		Matcher: matcher,
		Body:    body,
	}
	dbclass.Rules = append([]*JRule{rule}, dbclass.Rules...)
	for _, candidate := range dbclass.Facts {
		tryJRuleOn(rule, candidate)
	}
	return rule
}

// JTryRules tries all rules in the datum's dbclass against the datum.
// Corresponds to the Lisp function try-rules.
func JTryRules(datum *JDatum) {
	for _, rule := range datum.Dbclass.Rules {
		tryJRuleOn(rule, datum)
	}
}

// tryJRuleOn tries a single rule against a datum.
// If the matcher succeeds, the rule body is enqueued with the bindings.
// Corresponds to the Lisp function try-rule-on.
func tryJRuleOn(rule *JRule, datum *JDatum) {
	jtre := datum.Dbclass.Jtre
	ok, bindings, needsNode := rule.Matcher(datum.LispForm)
	if !ok {
		return
	}
	if needsNode {
		bindings = append([]interface{}{datum.TmsNode}, bindings...)
	}
	JtreEnqueue(&QueueEntry{Body: rule.Body, Bindings: bindings}, jtre)
}

// RunJRules executes all queued rules in the JTRE.
// Corresponds to the Lisp function run-rules.
func RunJRules(jtre *Jtre) {
	counter := 0
	for len(jtre.Queue) > 0 {
		entry := jtre.Queue[0]
		jtre.Queue = jtre.Queue[1:]
		entry.Body(entry.Bindings...)
		counter++
	}
	debuggingJtre(jtre, "\n    %d rules run.", counter)
	jtre.RulesRun += counter
}

// JRulesWaiting returns true if there are rules queued for execution.
// Corresponds to the Lisp function rules-waiting?.
func JRulesWaiting(jtre *Jtre) bool {
	return len(jtre.Queue) > 0
}

// JEnqueue adds a new entry to the JTRE's rule queue.
// Corresponds to the Lisp function enqueue.
func JEnqueue(entry *QueueEntry, jtre *Jtre) {
	jtre.Queue = append([]*QueueEntry{entry}, jtre.Queue...)
}

// JDequeue removes and returns the first entry from the queue.
// Corresponds to the Lisp function dequeue.
func JDequeue(jtre *Jtre) *QueueEntry {
	if len(jtre.Queue) == 0 {
		return nil
	}
	entry := jtre.Queue[0]
	jtre.Queue = jtre.Queue[1:]
	return entry
}

// ShowJRules displays all rules in the JTRE.
// Corresponds to the Lisp function show-rules.
func ShowJRules(jtre *Jtre, w io.Writer) {
	if w == nil {
		w = os.Stdout
	}
	fmt.Fprintf(w, "\nThere are %d rules in %s:", jtre.RuleCounter, jtre.Title)
	if len(jtre.Queue) == 0 {
		fmt.Fprintf(w, "\n None queued.")
	} else {
		fmt.Fprintf(w, "\n %d queued.", len(jtre.Queue))
	}
	JMapDbclass(func(dbclass *JDbclass) {
		for _, rule := range dbclass.Rules {
			PrintJRule(rule, w)
		}
	}, jtre)
}

// PrintJRule prints a single rule.
// Corresponds to the Lisp function print-rule.
func PrintJRule(rule *JRule, w io.Writer) {
	if w == nil {
		w = os.Stdout
	}
	fmt.Fprintf(w, "\n %s", rule)
}

// GetJRule finds a rule by its ID.
// Corresponds to the Lisp function get-rule.
func GetJRule(num int, jtre *Jtre) *JRule {
	var result *JRule
	JMapDbclass(func(dbclass *JDbclass) {
		for _, rule := range dbclass.Rules {
			if rule.ID == num {
				result = rule
			}
		}
	}, jtre)
	return result
}

File: jtms/java/JRules.java
// JTRE Rule system module.
// Translated from jrules.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.
package jtms;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

/**
 * JRules provides the rule system for JTRE:
 * inserting rules, trying rules against facts, and running the rule queue.
 */
public class JRules {

    /** Insert a new rule into a dbclass. Corresponds to insert-rule. */
    public static Rule insertRule(Dbclass dbclass, Rule.RuleMatcher matcher, Jtre.RuleBody body) {
        Jtre jtre = dbclass.getJtre();
        jtre.setRuleCounter(jtre.getRuleCounter() + 1);
        Rule rule = new Rule(jtre.getRuleCounter(), jtre, dbclass, matcher, body);
        dbclass.getRules().add(0, rule);
        for (Datum candidate : dbclass.getFacts()) {
            tryRuleOn(rule, candidate);
        }
        return rule;
    }

    /** Try all rules in a datum's dbclass against it. Corresponds to try-rules. */
    public static void tryRules(Datum datum) {
        for (Rule rule : datum.getDbclass().getRules()) {
            tryRuleOn(rule, datum);
        }
    }

    /** Try a single rule against a datum. Corresponds to try-rule-on. */
    public static void tryRuleOn(Rule rule, Datum datum) {
        Jtre jtre = datum.getDbclass().getJtre();
        Rule.MatchResult result = rule.getMatcher().match(datum.getLispForm());
        if (!result.ok) return;
        List<Object> bindings = result.bindings != null ?
            new ArrayList<>(result.bindings) : new ArrayList<>();
        if (result.needsNode) {
            bindings.add(0, datum.getTmsNode());
        }
        Jtre.enqueue(new Jtre.QueueEntry(rule.getBody(), bindings), jtre);
    }

    /** Run all queued rules. Corresponds to run-rules. */
    public static void runRules(Jtre jtre) {
        int counter = 0;
        while (!jtre.getQueue().isEmpty()) {
            Jtre.QueueEntry entry = jtre.getQueue().remove(0);
            entry.body.execute(entry.bindings);
            counter++;
        }
        jtre.debug("%n    %d rules run.", counter);
        jtre.setRulesRun(jtre.getRulesRun() + counter);
    }

    /** Check if rules are waiting. Corresponds to rules-waiting?. */
    public static boolean rulesWaiting(Jtre jtre) {
        return !jtre.getQueue().isEmpty();
    }

    /** Display all rules. Corresponds to show-rules. */
    public static void showRules(Jtre jtre, PrintStream out) {
        if (out == null) out = System.out;
        PrintStream finalOut = out;
        finalOut.printf("%nThere are %d rules in %s:", jtre.getRuleCounter(), jtre.getTitle());
        if (jtre.getQueue().isEmpty()) {
            finalOut.printf("%n None queued.");
        } else {
            finalOut.printf("%n %d queued.", jtre.getQueue().size());
        }
        JData.mapDbclass(dbclass -> {
            for (Rule rule : dbclass.getRules()) {
                finalOut.printf("%n %s", rule);
            }
        }, jtre);
    }

    /** Find a rule by ID. Corresponds to get-rule. */
    public static Rule getRule(int num, Jtre jtre) {
        final Rule[] result = {null};
        JData.mapDbclass(dbclass -> {
            for (Rule rule : dbclass.getRules()) {
                if (rule.getId() == num) { result[0] = rule; return; }
            }
        }, jtre);
        return result[0];
    }
}
